File: experiments/adrianco/experiment-14/runs/language=clojure_model=sonnet_prompt=neutral/rep1/test/brazilian_soccer_mcp/tools_test.clj
(ns brazilian-soccer-mcp.tools-test
  (:require [clojure.test :refer :all]
            [clojure.string :as str]
            [brazilian-soccer-mcp.normalize :as norm]
            [brazilian-soccer-mcp.data :as data]
            [brazilian-soccer-mcp.tools :as tools]))

;; ── Load data once for all tests ─────────────────────────────────────────────

(defonce _loaded (data/load-all-data!))

;; ── normalize tests ───────────────────────────────────────────────────────────

(deftest test-normalize-team
  (testing "strips state codes"
    (is (= "palmeiras" (norm/normalize-team "Palmeiras-SP")))
    (is (= "flamengo" (norm/normalize-team "Flamengo-RJ")))
    (is (= "sport" (norm/normalize-team "Sport-PE")))
    (is (= "america" (norm/normalize-team "América - MG"))))

  (testing "strips parenthetical notes"
    (is (str/includes? (norm/normalize-team "Boavista Sport Club (antigo Esporte Clube Barreira) - RJ")
                       "boavista")))

  (testing "strips accents"
    (is (= "gremio" (norm/normalize-team "Grêmio")))
    (is (= "atletico" (norm/normalize-team "Atlético")))
    (is (= "sao paulo" (norm/normalize-team "São Paulo"))))

  (testing "returns nil for nil/empty"
    (is (nil? (norm/normalize-team nil)))
    (is (nil? (norm/normalize-team "")))))

(deftest test-team-matches
  (testing "exact match after normalization"
    (is (norm/team-matches? "Palmeiras-SP" "Palmeiras"))
    (is (norm/team-matches? "Flamengo-RJ" "Flamengo"))
    (is (norm/team-matches? "Grêmio" "Gremio")))

  (testing "partial match"
    (is (norm/team-matches? "São Paulo FC" "Sao Paulo"))
    (is (norm/team-matches? "Corinthians" "corinthians")))

  (testing "no match"
    (is (not (norm/team-matches? "Palmeiras" "Flamengo")))
    (is (not (norm/team-matches? "Santos" "Grêmio"))))

  (testing "nil safety"
    (is (not (norm/team-matches? nil "Flamengo")))
    (is (not (norm/team-matches? "Flamengo" nil)))))

(deftest test-parse-date
  (is (= "2023-09-03" (norm/parse-date "2023-09-03")))
  (is (= "2012-05-19" (norm/parse-date "2012-05-19 18:30:00")))
  (is (= "2003-03-29" (norm/parse-date "29/03/2003")))
  (is (nil? (norm/parse-date nil)))
  (is (nil? (norm/parse-date ""))))

;; ── data loading tests ────────────────────────────────────────────────────────

(deftest test-data-loading
  (testing "brasileirao matches loaded"
    (let [ms (data/get-brasileirao)]
      (is (> (count ms) 4000))
      (is (every? :date (take 10 ms)))
      (is (every? :home-team (take 10 ms)))
      (is (every? :away-team (take 10 ms)))))

  (testing "copa brasil matches loaded"
    (let [ms (data/get-copa-brasil)]
      (is (> (count ms) 1000))))

  (testing "libertadores matches loaded"
    (let [ms (data/get-libertadores)]
      (is (> (count ms) 1000))))

  (testing "br-football extended dataset loaded"
    (let [ms (data/get-br-football)]
      (is (> (count ms) 5000))))

  (testing "historico dataset loaded"
    (let [ms (data/get-historico)]
      (is (> (count ms) 6000))))

  (testing "fifa players loaded"
    (let [ps (data/get-fifa)]
      (is (> (count ps) 10000))
      (is (every? :name (take 10 ps))))))

;; ── search_matches tests ──────────────────────────────────────────────────────

(deftest test-search-matches-by-team
  (testing "finds Flamengo matches"
    (let [result (tools/search-matches {"team" "Flamengo" "competition" "brasileirao" "limit" "50"})]
      (is (string? result))
      (is (str/includes? result "Flamengo"))
      (is (not (str/includes? result "No matches found")))))

  (testing "finds Palmeiras matches in 2022"
    (let [result (tools/search-matches {"team" "Palmeiras" "season" "2022" "competition" "brasileirao"})]
      (is (str/includes? result "2022"))
      (is (str/includes? result "Palmeiras"))))

  (testing "returns no-match message when nothing found"
    (let [result (tools/search-matches {"team" "NonExistentTeam12345"})]
      (is (str/includes? result "No matches found"))))

  (testing "respects limit"
    (let [result (tools/search-matches {"competition" "brasileirao" "limit" "5"})]
      ;; Should mention 5 or note total
      (is (string? result)))))

(deftest test-search-matches-head-to-head
  (testing "finds Flamengo vs Fluminense"
    (let [result (tools/search-matches {"team" "Flamengo" "opponent" "Fluminense"})]
      (is (str/includes? result "Flamengo"))
      (is (str/includes? result "Fluminense"))))

  (testing "finds Corinthians vs Palmeiras"
    (let [result (tools/search-matches {"team" "Corinthians" "opponent" "Palmeiras"
                                        "competition" "brasileirao"})]
      (is (string? result)))))

(deftest test-search-matches-by-date
  (testing "date range filter"
    (let [result (tools/search-matches {"date_from" "2023-01-01" "date_to" "2023-12-31"})]
      (is (string? result)))))

;; ── get_team_stats tests ──────────────────────────────────────────────────────

(deftest test-get-team-stats
  (testing "Flamengo 2019 stats"
    (let [result (tools/get-team-stats {"team" "Flamengo" "season" "2019"
                                        "competition" "brasileirao"})]
      (is (string? result))
      (is (str/includes? result "Flamengo"))
      (is (str/includes? result "Wins"))
      (is (str/includes? result "Win rate"))))

  (testing "missing team throws"
    (is (thrown? Exception (tools/get-team-stats {}))))

  (testing "unknown team returns no-data message"
    (let [result (tools/get-team-stats {"team" "XyzNonExistent"})]
      (is (str/includes? result "No matches found")))))

;; ── search_players tests ──────────────────────────────────────────────────────

(deftest test-search-players
  (testing "search by nationality Brazil"
    (let [result (tools/search-players {"nationality" "Brazil" "limit" "10"})]
      (is (str/includes? result "Brazil"))
      (is (not (str/includes? result "No players found")))))

  (testing "search by club Flamengo"
    (let [result (tools/search-players {"club" "Flamengo" "limit" "20"})]
      (is (string? result))))

  (testing "search by name"
    (let [result (tools/search-players {"name" "Neymar"})]
      (is (str/includes? result "Neymar"))))

  (testing "search by min overall rating"
    (let [result (tools/search-players {"min_overall" "90" "limit" "10"})]
      (is (string? result))
      (is (not (str/includes? result "No players found")))))

  (testing "search by position"
    (let [result (tools/search-players {"position" "GK" "nationality" "Brazil" "limit" "10"})]
      (is (string? result))))

  (testing "no match returns message"
    (let [result (tools/search-players {"name" "ZzzNobodyXxx123"})]
      (is (str/includes? result "No players found")))))

;; ── get_standings tests ───────────────────────────────────────────────────────

(deftest test-get-standings
  (testing "2019 brasileirao standings"
    (let [result (tools/get-standings {"season" "2019" "competition" "brasileirao"})]
      (is (string? result))
      (is (str/includes? result "2019"))
      (is (str/includes? result "Pts"))))

  (testing "missing season throws"
    (is (thrown? Exception (tools/get-standings {}))))

  (testing "copa brasil standings"
    (let [result (tools/get-standings {"season" "2020" "competition" "copa-brasil"})]
      (is (string? result)))))

;; ── get_head_to_head tests ────────────────────────────────────────────────────

(deftest test-get-head-to-head
  (testing "Flamengo vs Corinthians"
    (let [result (tools/get-head-to-head {"team1" "Flamengo" "team2" "Corinthians"})]
      (is (string? result))
      (is (str/includes? result "Head-to-Head"))
      (is (str/includes? result "Total matches"))))

  (testing "missing team throws"
    (is (thrown? Exception (tools/get-head-to-head {"team1" "Flamengo"}))))

  (testing "no results case"
    (let [result (tools/get-head-to-head {"team1" "TeamAAA" "team2" "TeamBBB"})]
      (is (str/includes? result "No matches found")))))

;; ── get_biggest_wins tests ────────────────────────────────────────────────────

(deftest test-get-biggest-wins
  (testing "overall biggest wins"
    (let [result (tools/get-biggest-wins {})]
      (is (string? result))
      (is (str/includes? result "Biggest victories"))))

  (testing "biggest wins for Palmeiras"
    (let [result (tools/get-biggest-wins {"team" "Palmeiras" "limit" "5"})]
      (is (string? result))))

  (testing "biggest wins in brasileirao 2019"
    (let [result (tools/get-biggest-wins {"competition" "brasileirao" "season" "2019" "limit" "5"})]
      (is (string? result)))))

;; ── get_competition_stats tests ───────────────────────────────────────────────

(deftest test-get-competition-stats
  (testing "brasileirao 2022 stats"
    (let [result (tools/get-competition-stats {"competition" "brasileirao" "season" "2022"})]
      (is (string? result))
      (is (str/includes? result "Goals per match"))
      (is (str/includes? result "Home wins"))))

  (testing "all competitions all years"
    (let [result (tools/get-competition-stats {})]
      (is (string? result))
      (is (str/includes? result "Total matches"))))

  (testing "libertadores stats"
    (let [result (tools/get-competition-stats {"competition" "libertadores" "season" "2019"})]
      (is (string? result)))))

;; ── integration / sample question tests ──────────────────────────────────────

(deftest test-sample-questions
  (testing "Q: When did Flamengo last play Corinthians?"
    (let [result (tools/search-matches {"team" "Flamengo" "opponent" "Corinthians" "limit" "1"})]
      (is (string? result))
      (is (not (str/includes? result "No matches found")))))

  (testing "Q: Who are the highest-rated players at Palmeiras?"
    (let [result (tools/search-players {"club" "Palmeiras" "limit" "5"})]
      (is (string? result))))

  (testing "Q: Who won the 2019 Brasileirao?"
    (let [result (tools/get-standings {"season" "2019" "competition" "brasileirao"})]
      (is (string? result))
      (is (str/includes? result "Flamengo"))))

  (testing "Q: What is the average goals per match in Brasileirao?"
    (let [result (tools/get-competition-stats {"competition" "brasileirao"})]
      (is (str/includes? result "Goals per match"))))

  (testing "Q: Find all Brazilian players"
    (let [result (tools/search-players {"nationality" "Brazil" "limit" "5"})]
      (is (str/includes? result "Brazil"))))

  (testing "Q: Find all Copa do Brasil finals"
    (let [result (tools/search-matches {"competition" "copa-brasil" "limit" "5"})]
      (is (string? result))))

  (testing "Q: Show biggest wins in Libertadores"
    (let [result (tools/get-biggest-wins {"competition" "libertadores" "limit" "5"})]
      (is (string? result))))

  (testing "Q: Corinthians home record in 2022"
    (let [result (tools/get-team-stats {"team" "Corinthians" "season" "2022"
                                        "competition" "brasileirao" "venue" "home"})]
      (is (string? result)))))

File: experiments/adrianco/experiment-55-terra-vs-opus5-effort/brazil/runs/agent=claude-code_effort=max_language=go_model=claude-opus-5_prompt=neutral/rep1/internal/soccerserver/tools_test.go
// tools_test.go covers the edges of each tool: bad arguments, ambiguous clubs,
// empty results and the answers that have to explain a gap in the data.
package soccerserver

import (
	"strings"
	"testing"

	"github.com/adriancockcroft/brazilian-soccer-mcp/internal/soccer"
)

func TestSearchMatchesFilters(t *testing.T) {
	data := structured(t, "search_matches", map[string]any{
		"team": "Flamengo", "opponent": "Fluminense", "competition": "brasileirao",
		"season_from": 2015, "season_to": 2019, "limit": 100,
	})
	matches := data["matches"].([]map[string]any)
	if len(matches) == 0 {
		t.Fatal("no Fla-Flu league matches between 2015 and 2019")
	}
	for _, m := range matches {
		if m["competition"] != soccer.CompBrasileirao {
			t.Errorf("competition filter leaked: %v", m["competition"])
		}
		if season := m["season"].(int); season < 2015 || season > 2019 {
			t.Errorf("season filter leaked: %d", season)
		}
	}
}

func TestSearchMatchesByDateRange(t *testing.T) {
	got := answer(t, "search_matches", map[string]any{
		"team": "Santos", "date_from": "2019-05-01", "date_to": "2019-05-31", "limit": 20})
	containsAll(t, got, "2019-05")
	if strings.Contains(got, "2019-06") || strings.Contains(got, "2019-04") {
		t.Errorf("date range leaked:\n%s", got)
	}
}

func TestSearchMatchesEmptyResultExplainsItself(t *testing.T) {
	got := answer(t, "search_matches", map[string]any{"team": "Flamengo", "competition": "serie-c"})
	containsAll(t, got, "No matches found", "Flamengo")
	if !strings.Contains(got, "appears in the data") {
		t.Errorf("an empty answer should say where the club does appear:\n%s", got)
	}
}

func TestAmbiguousTeamIsReported(t *testing.T) {
	got := failure(t, "search_matches", map[string]any{"team": "América"})
	containsAll(t, got, "could mean several clubs", "América-MG", "América-RN")
}

func TestUnknownTeamSuggests(t *testing.T) {
	got := failure(t, "team_stats", map[string]any{"team": "Palmieras"})
	if !strings.Contains(got, "Did you mean") && !strings.Contains(got, "Palmeiras") {
		t.Errorf("an unknown club should suggest alternatives:\n%s", got)
	}
}

func TestUnknownCompetitionIsReported(t *testing.T) {
	got := failure(t, "search_matches", map[string]any{"competition": "Premier League"})
	containsAll(t, got, "unknown competition", "brasileirao")
}

func TestUnknownSeasonIsReported(t *testing.T) {
	got := failure(t, "standings", map[string]any{"competition": "brasileirao", "season": 1975})
	containsAll(t, got, "available seasons", "2003")
}

func TestBadArgumentTypesAreRejected(t *testing.T) {
	cases := []struct {
		tool string
		args map[string]any
		want string
	}{
		{"search_matches", map[string]any{"season": "not a year"}, "integer"},
		{"search_matches", map[string]any{"date_from": "yesterday"}, "not a date"},
		{"search_matches", map[string]any{"team": "Flamengo", "date_from": "2019-12-01", "date_to": "2019-01-01"}, "before"},
		{"search_matches", map[string]any{"season_from": 2020, "season_to": 2010}, "before"},
		{"team_stats", map[string]any{"team": "Flamengo", "venue": "moon"}, "venue"},
		{"compare_seasons", map[string]any{"seasons": []any{2019}}, "at least two"},
	}
	for _, c := range cases {
		got := failure(t, c.tool, c.args)
		if !strings.Contains(strings.ToLower(got), strings.ToLower(c.want)) {
			t.Errorf("%s with %v: message %q does not mention %q", c.tool, c.args, got, c.want)
		}
	}
}

func TestHeadToHeadRejectsOneClubTwice(t *testing.T) {
	got := failure(t, "head_to_head", map[string]any{"team_a": "Flamengo", "team_b": "flamengo-rj"})
	containsAll(t, got, "two different clubs")
}

func TestMatchDetailsCarriesExtendedStats(t *testing.T) {
	got := answer(t, "match_details", map[string]any{
		"team": "Flamengo", "opponent": "Corinthians", "date": "2023-10-08"})
	containsAll(t, got, "2023-10-08", "Shots:", "corners:", "Sources:", "Match id:")
}

func TestMatchDetailsByID(t *testing.T) {
	data := structured(t, "search_matches", map[string]any{"team": "Flamengo", "season": 2019, "limit": 1})
	id := data["matches"].([]map[string]any)[0]["id"].(string)
	got := answer(t, "match_details", map[string]any{"match_id": id})
	containsAll(t, got, id)
	failure(t, "match_details", map[string]any{"match_id": "not-a-match"})
	failure(t, "match_details", map[string]any{"team": "Flamengo"})
}

func TestStandingsThroughRound(t *testing.T) {
	got := answer(t, "standings", map[string]any{"season": 2019, "through_round": 5, "limit": 5})
	containsAll(t, got, "2019 Brasileirão standings")
	if strings.Contains(got, "Champion") {
		t.Errorf("a five round table must not crown anybody:\n%s", got)
	}
}

func TestSeasonSummaryLeagueAndCup(t *testing.T) {
	league := answer(t, "season_summary", map[string]any{"competition": "brasileirao", "season": 2019})
	containsAll(t, league, "Campeonato Brasileiro Série A 2019", "Champion: Flamengo", "Season statistics", "Most goals scored")

	cup := answer(t, "season_summary", map[string]any{"competition": "copa-do-brasil", "season": 2019})
	containsAll(t, cup, "Bracket", "Final", "Winner")
}

func TestBiggestWinsOrdering(t *testing.T) {
	data := structured(t, "biggest_wins", map[string]any{"competition": "brasileirao", "limit": 5})
	matches := data["matches"].([]map[string]any)
	prev := 99
	for _, m := range matches {
		margin := m["home_goals"].(int) - m["away_goals"].(int)
		if margin < 0 {
			margin = -margin
		}
		if margin > prev {
			t.Error("biggest wins are not ordered by margin")
		}
		prev = margin
	}
	byGoals := answer(t, "biggest_wins", map[string]any{"order_by": "total_goals", "limit": 3})
	containsAll(t, byGoals, "Highest scoring matches")
}

func TestBestRecordsMetrics(t *testing.T) {
	for _, metric := range []string{"points", "win_rate", "wins", "goals_for", "goals_against", "goal_diff", "points_per_match"} {
		got := answer(t, "best_records", map[string]any{
			"competition": "brasileirao", "season": 2019, "metric": metric, "limit": 3})
		containsAll(t, got, strings.ReplaceAll(metric, "_", " "))
	}
}

func TestFindDerbies(t *testing.T) {
	got := answer(t, "find_derbies", map[string]any{"season": 2019, "competition": "brasileirao", "limit": 50})
	containsAll(t, got, "Derbies", "Fla-Flu")
	filtered := structured(t, "find_derbies", map[string]any{"rivalry": "Grenal", "limit": 10})
	for _, m := range filtered["matches"].([]map[string]any) {
		if m["rivalry"] != "Grenal" {
			t.Errorf("rivalry filter leaked: %v", m["rivalry"])
		}
	}
	none := answer(t, "find_derbies", map[string]any{"rivalry": "Not A Derby"})
	containsAll(t, none, "Known derbies")
}

func TestSearchPlayersMissingClubExplained(t *testing.T) {
	got := answer(t, "search_players", map[string]any{"club": "Flamengo"})
	containsAll(t, got, "No players match", "no squad for", "Grêmio")
}

func TestSearchPlayersByClubUsesTheCanonicalName(t *testing.T) {
	for _, club := range []string{"Gremio", "Grêmio", "gremio-rs"} {
		data := structured(t, "search_players", map[string]any{"club": club, "limit": 5})
		if data["total"].(int) == 0 {
			t.Errorf("no squad found for %q", club)
		}
	}
}

func TestClubSquadsFilters(t *testing.T) {
	all := structured(t, "club_squads", map[string]any{"brazilian_only": false, "limit": 200})
	brazilian := structured(t, "club_squads", map[string]any{"brazilian_only": true, "limit": 200})
	if brazilian["total"].(int) >= all["total"].(int) {
		t.Error("the Brazilian filter did not narrow the list")
	}
	named := answer(t, "club_squads", map[string]any{"clubs": []any{"Cruzeiro"}})
	containsAll(t, named, "Cruzeiro")
}

func TestGraphTools(t *testing.T) {
	got := answer(t, "graph_neighbors", map[string]any{"entity": "Flamengo", "type": "team", "limit": 10})
	containsAll(t, got, "team:flamengo-rj", "relationships")

	byID := answer(t, "graph_neighbors", map[string]any{
		"entity": "competition:brasileirao", "node_types": []any{"season"}, "limit": 5})
	containsAll(t, byID, "season")

	path := answer(t, "graph_path", map[string]any{
		"from": "Neymar", "from_type": "player", "to": "country:brazil", "max_depth": 3})
	containsAll(t, path, "Path from", "nationality")

	failure(t, "graph_neighbors", map[string]any{"entity": "not an entity at all"})

	disconnected := answer(t, "graph_path", map[string]any{
		"from": "team:flamengo-rj", "to": "competition:brasileirao", "max_depth": 1})
	containsAll(t, disconnected, "No path")
}

func TestListTeamsFilters(t *testing.T) {
	data := structured(t, "list_teams", map[string]any{"state": "RJ", "limit": 100})
	teams := data["teams"].([]map[string]any)
	if len(teams) < 5 {
		t.Fatalf("only %d clubs from Rio", len(teams))
	}
	for _, team := range teams {
		if team["state"] != "RJ" {
			t.Errorf("%v is not from RJ", team["name"])
		}
	}
	byComp := structured(t, "list_teams", map[string]any{"competition": "libertadores", "limit": 200})
	for _, team := range byComp["teams"].([]map[string]any) {
		found := false
		for _, comp := range team["competitions"].([]string) {
			if comp == soccer.CompLibertadores {
				found = true
			}
		}
		if !found {
			t.Errorf("%v never played the Libertadores", team["name"])
		}
	}
	bySeason := structured(t, "list_teams", map[string]any{"competition": "brasileirao", "season": 2019, "limit": 100})
	if n := len(bySeason["teams"].([]map[string]any)); n != 20 {
		t.Errorf("the 2019 Brasileirão had %d clubs", n)
	}
	byQuery := answer(t, "list_teams", map[string]any{"query": "atletico"})
	containsAll(t, byQuery, "Atlético-MG", "Atlético-GO")
}

func TestLimitsAreBounded(t *testing.T) {
	data := structured(t, "search_matches", map[string]any{"team": "Flamengo", "limit": 5000})
	if len(data["matches"].([]map[string]any)) > 200 {
		t.Error("limit is not capped")
	}
}

func TestDatasetInventory(t *testing.T) {
	got := answer(t, "list_datasets", map[string]any{})
	containsAll(t, got,
		"Brasileirao_Matches.csv", "novo_campeonato_brasileiro.csv", "Brazilian_Cup_Matches.csv",
		"Libertadores_Matches.csv", "BR-Football-Dataset.csv", "fifa_data.csv",
		"CC BY 4.0", "CC0 Public Domain", "Apache 2.0", "Known gaps")
}

func TestServerMetadata(t *testing.T) {
	srv := testServer(t)
	if srv.Name != ServerName || srv.Version != ServerVersion {
		t.Errorf("server identity = %s %s", srv.Name, srv.Version)
	}
	if !strings.Contains(srv.Instructions, "Coverage:") {
		t.Errorf("instructions should describe the coverage:\n%s", srv.Instructions)
	}
	if !strings.Contains(srv.Instructions, "Limitations") {
		t.Error("instructions should state the limitations")
	}
}

func TestVenueAndResultNeedAClub(t *testing.T) {
	containsAll(t, failure(t, "search_matches", map[string]any{"result": "win", "season": 2019}), "needs a `team`")
	containsAll(t, failure(t, "search_matches", map[string]any{"venue": "home", "season": 2019}), "needs a `team`")
}

func TestCompareSeasonsHandlesOverlappingArguments(t *testing.T) {
	got := answer(t, "compare_seasons", map[string]any{
		"competition": "brasileirao", "seasons": []any{2018, 2019},
		"season_from": 2018, "season_to": 2019,
	})
	if strings.Count(got, "\n2018 ") > 1 || strings.Count(got, "\n2019 ") > 1 {
		t.Errorf("a season is listed twice:\n%s", got)
	}
}

func TestClubSquadsDeduplicatesOverlappingPatterns(t *testing.T) {
	got := answer(t, "club_squads", map[string]any{"clubs": []any{"Gremio", "Gr"}})
	if strings.Count(got, "- Grêmio:") > 1 {
		t.Errorf("Grêmio is listed twice:\n%s", got)
	}
}

// A plain date_to means the whole of that day: fixtures carry a kick-off time.
func TestDateRangeIncludesTheLastDay(t *testing.T) {
	data := structured(t, "search_matches", map[string]any{
		"competition": "brasileirao", "date_from": "2019-05-26", "date_to": "2019-05-26", "limit": 20})
	matches := data["matches"].([]map[string]any)
	if len(matches) == 0 {
		t.Fatal("a single day range returned nothing, although fixtures were played that day")
	}
	for _, m := range matches {
		if m["date"] != "2019-05-26" {
			t.Errorf("date range leaked: %v", m["date"])
		}
	}
}

// Paging must be honest in the prose as well as in the structured payload.
func TestPagingProseMatchesTheData(t *testing.T) {
	all := structured(t, "search_matches", map[string]any{"team": "Flamengo", "season": 2019, "limit": 200})
	total := all["total"].(int)

	last := answer(t, "search_matches", map[string]any{
		"team": "Flamengo", "season": 2019, "limit": 3, "offset": total - 2})
	if strings.Contains(last, "more matches match these filters") {
		t.Errorf("the last page claims there is more to come:\n%s", last)
	}

	beyond := answer(t, "search_matches", map[string]any{
		"team": "Flamengo", "season": 2019, "limit": 3, "offset": total + 5})
	if strings.Contains(beyond, "No matches found") {
		t.Errorf("paging past the end must not read as an empty result set:\n%s", beyond)
	}
	containsAll(t, beyond, "past the end")

	players := answer(t, "search_players", map[string]any{"nationality": "Brazil", "limit": 2, "offset": 900})
	if strings.Contains(players, "No players match") {
		t.Errorf("paging past the end of the player list must explain itself:\n%s", players)
	}
}
